File: work_queue/src/bindings/perl/work_queue_example.pl
#! /usr/bin/env perl

# Copyright (c) 2010- The University of Notre Dame.
# This software is distributed under the GNU General Public License.
# See the file COPYING for details.
#
# This program is a very simple example of how to use Work Queue.
# It accepts a list of files on the command line.
# Each file is compressed with gzip and returned to the user.

use strict;
use warnings;

use Work_Queue;


# Main program:
unless (@ARGV) {
	die "work_queue_example <file1> [file2] [file3] ...\nEach file given on the command line will be compressed using a remote worker.\n";
}

# Usually, we can execute the gzip utility by simply typing its name at a
# terminal. However, this is not enough for work queue; we have to specify
# precisely which files need to be transmitted to the workers. We record the
# location of gzip in 'gzip_path', which is usually found in /bin/gzip or
# /usr/bin/gzip.
my $gzip_path = find_executable('gzip') || die "Could not find gzip anywhere in \$PATH";

# We create the tasks queue using the default port. If this port is already
# been used by another program, you can try setting port = 0 to use an
# available port.

my $port = $Work_Queue::WORK_QUEUE_DEFAULT_PORT;
my $q    = Work_Queue->new($port) || die "Instantiation of Work Queue failed! ($!)\n";

print "listening on port $port...\n";

# We create and dispatch a task for each filename given in the argument list
for my $infile (@ARGV) {
	my $outfile = $infile . ".gz";

	# Note that we write ./gzip here, to guarantee that the gzip
	# version we are using is the one being sent to the workers.
	my $command = "./gzip <$infile >$outfile";
	my $t = Work_Queue::Task->new($command);

	# gzip is the same across all tasks, so we can cache (the default)
	# it in the workers.  Note that when specifying a file, we have
	# to name its local name (e.g.  gzip_path), and its remote name
	# (e.g. "gzip"). Unlike the following line, more often than not
	# these are the same.
	$t->specify_input_file(local_name => $gzip_path, remote_name=>'gzip');

	# files to be compressed are different across all tasks, so we do
	# not cache them. This is, of course, application
	# specific. Sometimes you may want to cache an output file if is
	# the input of a later task. Note that remote_name defaults to
	# local_name when absent.
	$t->specify_input_file(local_name => $infile,   cache => $Work_Queue::WORK_QUEUE_NOCACHE);
	$t->specify_output_file(local_name => $outfile, cache => $Work_Queue::WORK_QUEUE_NOCACHE);

	# Once all files has been specified, we are ready to submit the
	# task to the queue.
	my $taskid = $q->submit($t);
	print "submitted task (id# ",  $t->id, '): ', $t->command, "\n";
}

print "waiting for tasks to complete...\n";

while (not $q->empty) {
	my $t = $q->wait(5);

	if($t) {
	print 'task (id# ', $t->id, ') complete: ', $t->command, '(return code ', $t->return_status, ")\n";

	if($t->return_status != 0) {
		# Task failed. Error handling here.
	}
	}
}

print "all tasks complete!\n";

sub find_executable {
	my $executable = shift;

	my @dirs = split ':', $ENV{PATH};
	push @dirs, '.';

	my @paths = map { "$_/$executable" } @dirs;

	for my $path (@paths) {
	return $path if -x $path;
	}

	undef;
}

File: work_queue/src/bindings/python3/work_queue_example.py
#!/usr/bin/env python

# Copyright (c) 2021- The University of Notre Dame.
# This software is distributed under the GNU General Public License.
# See the file COPYING for details.

# This program is a simple example of how to use Work Queue.
# It estimates the value of pi by estimating the ratio of the areas of a circle
# inscribed in a square.
# It accepts two parameters: number of tasks and number of random points to generate per tasks.
# Each task generates its own estimate independently, which then gets combined
# with other tasks as they finish.

import work_queue as wq
import sys
import os


def show_help():
    print("""Estimate pi using random points.

    Usage: work_queue_basic_example.py NUM_TASKS NUM_POINTS
    where:
       NUM_TASKS   Number of tasks to generate.
       NUM_POINTS  Number of random points to generate per task.

Each task generates its own pi etimate by generating NUM_POINTS random points.
These estimates are combined among tasks as tasks finish.""")


def read_arguments():
    try:
        (num_of_tasks, points_per_task) = sys.argv[1:]
    except ValueError:
        show_help()
        print("Error: Incorrect number of arguments")
        exit(1)

    try:
        num_of_tasks = int(num_of_tasks)
        points_per_task = int(points_per_task)
    except ValueError:
        print("Error: Arguments are not integers")
    return (num_of_tasks, points_per_task)


def read_results_file(task):
    result_file = "task_{}.out".format(task.tag)
    try:
        with open(result_file) as f:
            (points_in_square, points_in_circle, pi_estimate) = (float(x) for x in f.readline().split())
            return (points_in_square, points_in_circle)
    except Exception as e:
        print("Error reading file: {}: {}".format(result_file, e))
        return (0, 0)

def cleanup_task(task):
    result_file = "task_{}.out".format(task.tag)
    error_log = "task_{}.err".format(task.tag)

    for filename in [result_file, error_log]:
        try:
            os.remove(filename)
        except Exception as e:
            print("Could not remove task file: {}".format(filename))

def print_error_log(task_tag):
    error_log = "task_{}.err".format(task_tag)
    try:
        with open(error_log) as f:
            print(f.read())
    except IOError:
        pass


(num_of_tasks, points_per_task) = read_arguments()


q = wq.WorkQueue(9123)

for task_tag in range(num_of_tasks):
    output_name = "task_{}.out".format(task_tag)
    error_name  = "task_{}.err".format(task_tag)

    t = wq.Task("./area_circ_sq.py {num_points} {out} 2> {err}".format(num_points=points_per_task, out=output_name, err=error_name))

    t.specify_tag("{}".format(task_tag))

    t.specify_input_file("area_circ_sq.py", cache=True)
    t.specify_output_file(output_name, cache=False)
    t.specify_output_file(error_name, cache=False)

    t.specify_cores(1)
    t.specify_memory(100)
    t.specify_disk(200)

    print("Submitting task {}".format(task_tag))
    q.submit(t)


total_points_in_square = 0
total_points_in_circle = 0

# wait for all tasks to complete
while not q.empty():
    t = q.wait(5)
    if t:
        print("Task {id} has returned.".format(id=t.id))
        if t.result == wq.WORK_QUEUE_RESULT_SUCCESS and t.return_status == 0:
                points_in_square, points_in_circle = read_results_file(t)
                total_points_in_square += points_in_square
                total_points_in_circle += points_in_circle
                print("New estimate: {}".format(float(4 * total_points_in_circle) / total_points_in_square))
        else:
            print("    It could not be completed successfully. Result: {}. Exit code: {}".format(t.result, t.return_status))
            print_error_log(task_tag)
        cleanup_task(t)


print("All task have finished.")

